File: hub/sdk/node-client/src/discovery.ts
import { Config } from "./config.js";
import { HubError } from "./errors.js";

export interface DiscoveryDoc {
  issuer: string;
  authorization_endpoint: string;
  token_endpoint: string;
  userinfo_endpoint: string;
  jwks_uri: string;
  introspection_endpoint?: string;
  scopes_supported: string[];
  response_types_supported: string[];
  id_token_signing_alg_values_supported: string[];
  code_challenge_methods_supported: string[];
  [k: string]: unknown;
}

/** Fetch + cache /.well-known/openid-configuration. */
export class Discovery {
  private cached: DiscoveryDoc | null = null;
  private cachedAt = 0;

  constructor(private readonly config: Config) {}

  async get(): Promise<DiscoveryDoc> {
    const now = Date.now();
    if (this.cached && (now - this.cachedAt) / 1000 < this.config.jwksCacheTtl) {
      return this.cached;
    }
    const url = `${this.config.hubUrl}/.well-known/openid-configuration`;
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), this.config.httpTimeout);
    try {
      const r = await fetch(url, { signal: ctrl.signal, headers: { Accept: "application/json" } });
      if (!r.ok) throw new HubError(`Discovery HTTP ${r.status}`);
      const doc = (await r.json()) as DiscoveryDoc;
      if (!doc.issuer || !doc.jwks_uri) throw new HubError("Discovery missing required fields");
      this.cached = doc;
      this.cachedAt = now;
      return doc;
    } catch (e) {
      if (e instanceof HubError) throw e;
      throw new HubError(`Discovery fetch failed: ${(e as Error).message}`);
    } finally {
      clearTimeout(t);
    }
  }
}

File: hub/sdk/python-client/src/central_auth_hub/discovery.py
"""OIDC Discovery fetch + memory cache."""

from __future__ import annotations

import time

import httpx

from .config import Config
from .errors import HubError


class Discovery:
    """Fetch /.well-known/openid-configuration with TTL cache.

    Cache lives in instance — for cross-request cache, share instance.
    """

    def __init__(self, config: Config):
        self.config = config
        self._cached: dict | None = None
        self._cached_at: float = 0.0

    def get(self) -> dict:
        now = time.time()
        if self._cached and (now - self._cached_at) < self.config.jwks_cache_ttl:
            return self._cached

        url = f"{self.config.hub_url}/.well-known/openid-configuration"
        try:
            r = httpx.get(url, timeout=self.config.http_timeout)
            r.raise_for_status()
            doc = r.json()
        except Exception as e:
            raise HubError(f"Discovery fetch failed: {e}") from e

        if "issuer" not in doc or "jwks_uri" not in doc:
            raise HubError("Discovery missing required fields")
        self._cached = doc
        self._cached_at = now
        return doc

    async def get_async(self) -> dict:
        now = time.time()
        if self._cached and (now - self._cached_at) < self.config.jwks_cache_ttl:
            return self._cached
        url = f"{self.config.hub_url}/.well-known/openid-configuration"
        try:
            async with httpx.AsyncClient(timeout=self.config.http_timeout) as client:
                r = await client.get(url)
                r.raise_for_status()
                doc = r.json()
        except Exception as e:
            raise HubError(f"Discovery fetch failed: {e}") from e
        if "issuer" not in doc or "jwks_uri" not in doc:
            raise HubError("Discovery missing required fields")
        self._cached = doc
        self._cached_at = now
        return doc
